### ISSUE TITLE
hint for your questions

### ISSUE THREAD

**Issue by Nightbr** [Grafikart/BlogMVC-Symfony3#1]
Hi, 
thank you for this project, and we can't wait for all your Symfony tutorials ;)

I work with Symfony everyday so I'm happy to help you by trying to answering as best as possible to your questions:

*To "counter cache" the number of posts associated to each Category I created an EventSubscriber but I have to check the entity type. Is there a way to attach a subscriber to an Entity using annotation or anything else ?*

You can create method on Post entity with the annotation `@ORM\PreUpdate`. E.g.

```
/**
 * @ORM\PreUpdate
 */
public function preUpdate(PreUpdateEventArgs $event)
{
       if ($event->hasChangedField('category')) {
                $event
                    ->getEntityManager()
                    ->getRepository("AppBundle:Category")
                    ->decrementCount($event->getOldValue('category'))
                    ->incrementCount($event->getNewValue('category'));
         }
}
```

It is linked to Doctrine behavior: <http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/events.html#lifecycle-events>

*To avoid n+1 queries on a ManyToOne I have to give up the findAll / findBy methods and use the queryBuilder to make a LEFT JOIN. Is there a better way ? Does it affect Doctrine hydratation ?*

There is a better solution. You can change **Doctrine association fetch mode**. By default, association fetch mode is **LAZY**. It will calculate and fetch later the related entities if needed. You can change this fetch mode to **EAGER**, it will force fetch the related entities.

```
   /**
     * @ORM\ManyToOne(targetEntity="User", inversedBy="posts", fetch="EAGER")
     * @ORM\JoinColumn(name="user_id", referencedColumnName="id")
     */
    private $user;
```

<http://stackoverflow.com/questions/26891658/what-is-the-difference-between-fetch-eager-and-fetch-lazy-in-doctrine>

*I split my backend in a bundle instead of namespace. is it considered a good or bad practice ? Is there a way to prefix all routes inside a Bundle ?*

It's not a good or a bad practice. For example, in my last project, I have an `Admin` folder in each bundle with my Admin class and it works well. But when you begin to have a lot of bundle or logic link to your admin panel, it's good to separate it into an AdminBundle. I think a separate bundle is the most scalable option.

For prefix all routes inside a Bundle you have already done this:

```
admin:
    resource: "@AdminBundle/Controller/"
    type:     annotation
    prefix: /admin
```

But if you mean prefixing route name, it is more complicated (e.g. $myprefix_post_index). The only way is a bit complicated... I have seen this in the FosRestBundle and they generate the routing automatically:
<https://github.com/FriendsOfSymfony/FOSRestBundle/tree/master/Routing> 

*The sidebar need to be shared accress pages. I created a service injected as a "global" twig variable PartialService. Is there a better way to achieve that ?*

It seems to be good. For this, I used to work with render controller: 
<https://symfony.com/doc/current/templating/embedding_controllers.html>

*For the timestamps (created_at and updated_at) I used a trait since it seemed to be the easies way to reach the goal. Is it considered bad*

Use a trait is a really best practice since it is support by PHP. For example, gedmo DoctrineExtension create a timestampable trait for this behavior (and I use mine too in my project).
<http://apigen.juzna.cz/doc/l3pp4rd/DoctrineExtensions/class-Gedmo.Timestampable.Traits.TimestampableEntity.html>

That's it!

Thanks for your work and feel free to contact me if you need any more help!

**Comment by Grafikart** [comment]
Thanks for the help, I still have some concern about the first 2 answers.

- For the preUpdate Hook I think i'll stick with the EventSubscriber to keep database logic out of the entity. 
- I found the "EAGER" fetch somewhere too. This eager fetch will affect every query even if I don't need the category :(

**Comment by Nightbr** [comment]
Fair enough ;) So for the **EventSubscriber**, it's a Doctrine Event and it's a real mess... In fact, this event is triggered each time an entity is preupdate, ... That's why you need to verify the instanceof...

Yep for the *EAGER* it affects all queries. For example, I use this for associations such as user.avatar or when I am sure I will need the associate entity everytime. You gain some ms in dbquery. But the LAZY mode is the best in the main case. Also, you can use **EXTRA_LAZY** mode which is specifically made for associations with huge data. You can get only a specific part of the collection.

<http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/tutorials/extra-lazy-associations.html>

**Comment by lobnaKh** [comment]
Hello, What are the credentials of the backend please ??

**Comment by Grafikart** [comment]
@lobnaKh admin / admin 
https://github.com/Grafikart/BlogMVC-Symfony3/blob/master/src/AppBundle/DataFixtures/ORM/LoadUserData.php#L30
